File: app/src/main/java/com/jugger/tetris/Block/BlockOperations.java
package com.jugger.tetris.Block;


import com.jugger.tetris.Basic.Brick;
import com.jugger.tetris.Basic.C;

public class BlockOperations {

    public static Block generateNewRandomBlock() {
        return new Block(BlockTypes.getRandomBlockType(),-1);
    }

    public static Block generateNewRandomBlock_withImages() {
        int x = (int) (Math.random()* C.images.getCount());
        return new Block(BlockTypes.getRandomBlockType(),x);
    }



    public static Block getBlockRotation(Block b) {
        Brick[][] oldC = b.getComposition();
        Brick[][] newC = new Brick[4][4];
        Brick[] singleBricks = new Brick[4];
        int i =0;
        for(int r = 0; r < 4;r++) {
            for(int c = 0; c < 4;c++) {
                if(oldC[r][c] != null) {
                    newC[c][3-r] = oldC[r][c].copy();
                    newC[c][3-r].changePos(-c+r, -c+3-r);
                    singleBricks[i] = newC[c][3-r];
                    i++;
                }
            }
        }

        return new Block(singleBricks,newC,b.getBlockType());
    }

    public static void printBlockToConsole(Block b) {
        Brick[][] bricks = b.getComposition();
        System.out.println("Blockperspective!");
        for(int r = 0; r < 4;r++) {
            for(int c = 0; c < 4;c++) {
                if(bricks[r][c] != null) {
                    System.out.print("X");
                } else {
                    System.out.print("-");
                }
            }
            System.out.println();
        }
        System.out.println("____________");
    }

    public static Block moveBlock_downByOne(Block b) {
        Brick[] bricks = b.getBricks();
        for(int i = 0; i < bricks.length;i++) {
            bricks[i].changePos(-1, 0);
        }
        return b;
    }

    public static Block moveBlock_rightByOne(Block b) {
        Brick[] bricks = b.getBricks();
        for(int i = 0; i < bricks.length;i++) {
            bricks[i].changePos(0, +1);
        }
        return b;
    }

    public static Block moveBlock_leftByOne(Block b) {
        Brick[] bricks = b.getBricks();
        for(int i = 0; i < bricks.length;i++) {
            bricks[i].changePos(0, -1);
        }
        return b;
    }


}

File: app/src/main/java/com/jugger/tetris/Block/BlockOperations.kt
package com.jugger.tetris.Block

import com.jugger.tetris.Basic.Brick
import com.jugger.tetris.Constants

object BlockOperations {
    fun generateNewRandomBlock(): Block {
        return Block(BlockTypes.randomBlockType, -1)
    }

    fun generateNewRandomBlock_withImages(): Block {
        val x = (Math.random() * Constants.images.count).toInt()
        return Block(BlockTypes.randomBlockType, x)
    }

    fun getBlockRotation(b: Block): Block {
        val oldC = b.composition
        val newC = Array(4) { arrayOfNulls<Brick>(4) }
        val singleBricks = mutableListOf<Brick>()
        for (r in 0..3) {
            for (c in 0..3) {
                if (oldC[r][c] != null) {
                    newC[c][3 - r] = oldC[r][c]!!.copy()
                    newC[c][3 - r]!!.changePos(-c + r, -c + 3 - r)
                    singleBricks.add(newC[c][3 - r]!!)
                }
            }
        }
        return Block(singleBricks, newC, b.blockType)
    }


    fun moveDown(block: Block): Block {
        val b =block.copy()
        val bricks = b.bricks
        for (i in bricks.indices) {
            bricks[i].changePos(-1, 0)
        }
        return b
    }

    fun moveRight(block: Block): Block {
        val b =block.copy()
        val bricks = b.bricks
        for (i in bricks.indices) {
            bricks[i].changePos(0, +1)
        }
        return b
    }

    fun moveLeft(block: Block): Block {
        val b =block.copy()
        val bricks = b.bricks
        for (i in bricks.indices) {
            bricks[i].changePos(0, -1)
        }
        return b
    }
}